### DISCUSSION TITLE
some questions about objective function (mterm and/or lterm)

### DISCUSSION THREAD

**Discussion by zmdntnu** [do-mpc/do-mpc#228] [Q&A] [answered]
Dear @4flixt,

Hi, I wonder if there is a way to prioritize one of the objective when the objective function consists of several equations representing different optimization targets. I tried to use weights in my definition, but it seems that it's not so effective, and very often it fails to achieve the optimization goal.  This is how I define my objective function (get to appointed point and lower the risk or to avoid collision):

mterm =  (N- 400)**2 + (E - 600)**2 + 0.1*(model.aux['CRI']**2)
lterm = mterm

Second question is that if it's possible to set up conditional objective functions, for instance, adding a activation function so that the optimization target can vary along with the requirement. Say when activation function is not activated, then it tries to find the way from point A to B, and once the activation function is on, then it not only needs to find the path but also avoid the danger(moving obstacle).

I don't know if I've made my questions clear. Please let me know if there's clarification needed.

Best wishes

zmdntnu

**Comment by 4flixt** [comment, marked as answer]
Hi @zmdntnu, 

I again strongly advise you to read more about MPC and optimization before you start working with do-mpc. The tool is developed for people with experience in both fields and this forum is not meant to teach about fundamentals or to aid you with your personal projects.
However, as before, a brief answer:

1) Weighting is the way to go to prioritize competing objectives
2) Conditional objective functions (e.g. with CasADi ``if_else``)  are not recommended. These functions lead to problems for the optimizer due to them not being differentiable. 

Best,
4flixt
